cocotb/drivers/amba.py

```diff
@@ -26,31 +26,43 @@
 
 """Drivers for Advanced Microcontroller Bus Architecture."""
 
+import array
+from typing import Any, List, Optional, Sequence, Tuple, Union
+
 import cocotb
-from cocotb.triggers import RisingEdge, ReadOnly, Lock
-from cocotb.drivers import BusDriver
 from cocotb.binary import BinaryValue
-
-import array
+from cocotb.drivers import BusDriver
+from cocotb.handle import SimHandleBase
+from cocotb.result import TestError
+from cocotb.triggers import Combine, Lock, ReadOnly, RisingEdge
 
 
 class AXIProtocolError(Exception):
     pass
 
 
-class AXI4LiteMaster(BusDriver):
-    """AXI4-Lite Master.
+class AXI4Master(BusDriver):
+    """AXI4 Master
 
     TODO: Kill all pending transactions if reset is asserted.
     """
 
-    _signals = ["AWVALID", "AWADDR", "AWREADY",        # Write address channel
-                "WVALID", "WREADY", "WDATA", "WSTRB",  # Write data channel
-                "BVALID", "BREADY", "BRESP",           # Write response channel
-                "ARVALID", "ARADDR", "ARREADY",        # Read address channel
-                "RVALID", "RREADY", "RRESP", "RDATA"]  # Read data channel
+    _signals = \
+        ["AWVALID", "AWADDR", "AWREADY", "AWID", "AWLEN", "AWSIZE", "AWBURST",
+         "WVALID", "WREADY", "WDATA", "WSTRB",
+         "BVALID", "BREADY", "BRESP", "BID",
+         "ARVALID", "ARADDR", "ARREADY", "ARID", "ARLEN", "ARSIZE", "ARBURST",
+         "RVALID", "RREADY", "RRESP", "RDATA", "RID", "RLAST"]
+
+    _optional_signals = ["AWREGION", "AWLOCK", "AWCACHE", "AWPROT", "AWQOS",
+                         "WLAST",
+                         "ARREGION", "ARLOCK", "ARCACHE", "ARPROT", "ARQOS"]
 
-    def __init__(self, entity, name, clock, **kwargs):
+    _xRESP_msg = ("OKAY", "EXOKAY", "SLVERR", "DECERR")
+    _BURST_values = {"FIXED": 0b00, "INCR": 0b01, "WRAP": 0b10}
+
+    def __init__(self, entity: SimHandleBase, name: str, clock: SimHandleBase,
+                 **kwargs: Any):
```

Comment:
If you're willing to split this PR in two, we'd be more than happy to merge the type annotations by themselves.